File: Depth-First Search/Combinations/String/LET_no291_Word Pattern II.py
class Solution:
    def wordPatternMatch(self, pattern: str, s: str) -> bool:
        return self.is_match(pattern, 0, s, 0, {}, set())

    def is_match(self, s1, i, s2, j, memo, used_words):
        if i == len(s1):
            return j == len(s2)
        
        c = s1[i]
        if c in memo:
            word = memo[c]
            if not s2[j:].startswith(word):
                return False
            return self.is_match(s1, i + 1, s2, j + len(word), memo, used_words)

        for j2 in range(j + 1, len(s2) + 1):
            word = s2[j:j2]
            if word not in used_words:
                used_words.add(word)
                memo[c] = word
                if self.is_match(s1, i + 1, s2, j2, memo, used_words):
                    return True
                del memo[c]
                used_words.remove(word)
        
        return False

File: Depth-First Search/Combinations/String/LET_no291_Word Pattern II.cpp
class Solution {
public:
    bool wordPatternMatch(string pattern, string s) {
        unordered_map<char, string> ch2words;
        unordered_set<string> used_words;
        return helper(pattern, 0, s, 0, ch2words, used_words);
    }

    bool helper(string& pattern, int i, string& s, int j, unordered_map<char, string>& ch2words, unordered_set<string>& used_words) {
        if (i == pattern.size())
            return j == s.size();

        char c = pattern[i];
        auto it = ch2words.find(c);
        if (it != ch2words.end()) {
            string str = it->second;
            string::size_type n = s.find(str, j);
            if (n == string::npos)
                return false;
            return helper(pattern, i + 1, s, j + str.size(), ch2words, used_words);
        }

        int n = s.size();
        for (int len = 1; len <= n - j; len++) {
            string str = s.substr(j, len);
            if (!used_words.count(str)) {
                ch2words[c] = str;
                used_words.insert(str);
                if (helper(pattern, i + 1, s, j + len, ch2words, used_words))
                    return true;
                used_words.erase(str);
                ch2words.erase(c);
            }
        }

        return false;
    }
};